### PR TITLE
[FIRRTL][Dedup] Directly use SHA256 byte array instead of converting to string.

### PR DESCRIPTION
**Pull request by richardxia**
I broke the [Windows build](https://github.com/llvm/circt/runs/6015600867?check_suite_focus=true) in https://github.com/llvm/circt/pull/2886, due to some type casting that MVSC doesn't like.

In that LLVM submodule bump, LLVM changed the SHA256 hashers to return
std::array<uint8_t, 32> instead of StringRefs. I attempted to avoid
dealing with this in FIRRTL's Dedup pass by explicitly converting the
array back into a string, but the way the casts were being done was
causing issues with MSVC.

Changing the code to work directly with the std::array is more direct
and efficient, and obviates the issues with casting.

This adds a custom implementation for the DenseMapInfo trait
specifically for std::array<uint8_t, 32>, which allows us to directly
use the SHA256 hash results as map keys.

I should mention that this goes way beyond my normal comfort zone in C++, since I've never written a trait implementation before. I'm also not sure what the best and safest way to set the empty key, the tombstone key, and convert the 32-byte hash array into a single `unsigned`, so I'm definitely looking to get some feedback on that.

I manually kicked off a Windows test build at https://github.com/llvm/circt/runs/6017675451?check_suite_focus=true, although it's on a slightly older version of this branch, but since it takes so long to run, I figured I might as well submit this now to get feedback on this.

### CODE REVIEW

**Comment by richardxia** [comment]
This doesn't fix the Windows build error because I realized that it's actually [a totally different set of lines](https://github.com/llvm/circt/blob/654d5e96c84c80274f04ddd1a5984adc547b5cec/lib/Dialect/FIRRTL/Transforms/Dedup.cpp#L56-L61) that had the erroneous cast. I'm going to work on a separate fix for the actual error, but I think the change in this PR is still generally useful, so I'll keep it open.

File: lib/Dialect/FIRRTL/Transforms/Dedup.cpp [hunk 1]
```diff
@@ -856,6 +857,35 @@ void fixupAllModules(InstanceGraph &instanceGraph) {
...
+    std::array<uint8_t, 32> key;
+    std::fill(key.begin(), key.end(), ~0 - 1);
+    return key;
+  }
+
+  static unsigned getHashValue(const std::array<uint8_t, 32> &val) {
+    // We assume SHA256 is already a good hash and just truncate down to the
+    // number of bytes we need for DenseMap.
+    unsigned hash;
+    std::memcpy(&hash, &val, sizeof(unsigned));
```
Comment (darthscsi) [inline lib/Dialect/FIRRTL/Transforms/Dedup.cpp:879]: &val -> val.data()

**Review by darthscsi — APPROVED** [review]
LGTM.  Can tomestone and empty be valid data values?  If not the key might need to be 33 bytes.

**Comment by richardxia** [comment]
> LGTM. Can tomestone and empty be valid data values? If not the key might need to be 33 bytes.

As far as I can tell, yes, then can be valid data values, or at least the [existing DenseMapInfo implementations](https://github.com/llvm/llvm-project/blob/f14ebe91c5dd6be5b64a45e479291cd08676be0c/llvm/include/llvm/ADT/DenseMapInfo.h#L91-L92) don't do anything to stop you from colliding with the empty and tombstone values. For the integer types, they seem to always choose the highest and second highest possible positive values (`~0U` for unsigned types and `0x7FFF...` for signed types), and I guess they are assuming that the user of the DenseMap won't have keys with those values.

I think we're possibly even safer here, regardless of which key values we choose, since we're using a hash value which should in theory have a low chance of collision.

File: lib/Dialect/FIRRTL/Transforms/Dedup.cpp (hunk 1 above)
Comment (richardxia) [reply lib/Dialect/FIRRTL/Transforms/Dedup.cpp]: Thanks, fixed in an amended commit. Having made this mistake two PRs in a row, I have now learned to always use `.data()` and stop assuming that `std::array` data can be safely gotten from its pointer address.

**Pull request merged**